[SPARK-33440][CORE] Use current timestamp with warning log in HadoopFSDelegationTokenProvider when the issue date for token is not set up properly

HeartSaVioR opened the pull request:
### What changes were proposed in this pull request?

This PR proposes to use current timestamp with warning log when the issue date for token is not set up properly. The next section will explain the rationalization with details.

### Why are the changes needed?

Unfortunately not every implementations respect the `issue date` in `AbstractDelegationTokenIdentifier`, which Spark relies on while calculating. The default value of issue date is 0L, which is far from actual issue date, breaking logic on calculating next renewal date under some circumstance, leading to 0 interval (immediate) on rescheduling token renewal.

In HadoopFSDelegationTokenProvider, Spark calculates token renewal interval as below:

https://github.com/apache/spark/blob/2c64b731ae6a976b0d75a95901db849b4a0e2393/core/src/main/scala/org/apache/spark/deploy/security/HadoopFSDelegationTokenProvider.scala#L123-L134

The interval is calculated as `token.renew() - identifier.getIssueDate`, which is providing correct interval assuming both `token.renew()` and `identifier.getIssueDate` produce correct value, but it's going to be weird when `identifier.getIssueDate` provides 0L (default value), like below:

```
20/10/13 06:34:19 INFO security.HadoopFSDelegationTokenProvider: Renewal interval is 1603175657000 for token S3ADelegationToken/IDBroker
20/10/13 06:34:19 INFO security.HadoopFSDelegationTokenProvider: Renewal interval is 86400048 for token HDFS_DELEGATION_TOKEN
```

Hopefully we pick the minimum value as safety guard (so in this case, `86400048` is being picked up), but the safety guard leads unintentional bad impact on this case.

https://github.com/apache/spark/blob/2c64b731ae6a976b0d75a95901db849b4a0e2393/core/src/main/scala/org/apache/spark/deploy/security/HadoopFSDelegationTokenProvider.scala#L58-L71

Spark leverages the interval being calculated in above, "minimum" value of intervals, and blindly adds the value to token's issue date to calculates the next renewal date for the token, and picks "minimum" value again. In problematic case, the value would be `86400048` (86400048 + 0) which is quite smaller than current timestamp.

https://github.com/apache/spark/blob/2c64b731ae6a976b0d75a95901db849b4a0e2393/core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala#L228-L234

The next renewal date is subtracted with current timestamp again to get the interval, and multiplexed by configured ratio to produce the final schedule interval. In problematic case, this value goes to negative.

https://github.com/apache/spark/blob/2c64b731ae6a976b0d75a95901db849b4a0e2393/core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala#L180-L188

There's a safety guard to not allow negative value, but that's simply 0 meaning schedule immediately. This triggers next calculation of next renewal date to calculate the schedule interval, lead to the same behavior, hence updating delegation token immediately and continuously.

As we fetch token just before the calculation happens, the actual issue date is likely slightly before, hence it's not that dangerous to use current timestamp as issue date for the token the issue date has not been set up properly. Still, it's better not to leave the token implementation as it is, so we log warn message to let end users consult with token implementer.

### Does this PR introduce _any_ user-facing change?

Yes. End users won't encounter the tight loop of schedule of token renewal after the PR. In end users' perspective of reflection, there's nothing end users need to change.

### How was this patch tested?

Manually tested with problematic environment.

HeartSaVioR committed:
[SPARK-33440][CORE] Separate the calculation for the next renewal date for each delegation token

SparkQA commented:
**[Test build #131036 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131036/testReport)** for PR 30366 at commit [`38a6c5a`](https://github.com/apache/spark/commit/38a6c5a7232de90f41cc00342173ae41884fb9ad).

HeartSaVioR commented:
I sought into the test code to add the test, but I haven't found any on calculation. 

I'll need to mock HadoopFSDelegationTokenProvider in any way (mocking, subclassing) which also requires changing some methods from `private` to at least `private[deploy]`. I'll need to also deal with Credentials and create some Tokens & AbstractDelegationTokenIdentifiers for test purpose. Given the actual change is 50 lines, the requirements of the testing sounds like a bit huge, but I'll try out if we think it's still necessary to construct the test.

HeartSaVioR commented:
cc. @vanzin @squito @tgravescs @jerryshao Appreciate the reviewing. Thanks in advance!

HeartSaVioR commented:
Oh also cc. @gaborgsomogyi as he worked on this part before.

HeartSaVioR commented:
Another possible improvement (though I'm not 100% sure about the risk) would be considering `issue date` as current timestamp when the value is 0L (default). This might calculate the interval smaller than the actual (hence the risk), but as we fetch the token before doing the calculation (please correct me if I'm missing here), the alternative won't be too skewed.

This improvement is also helpful to address another possible issue where the problematic token identifier is solely used. Spark will correctly calculates the next renewal date (min interval is actually expired date so it's going to be correct after adding 0L), but Spark also caches the interval which is "unchanged", and adding 0L to the cached interval won't change the next renewal date, hence eventually it will be earlier than current timestamp, triggering the same problem.

Please review the additional possible improvement as well. I'll incorporate it once reviewers double-check about the idea and say it's OK to go.

HeartSaVioR commented:
Btw, I guess, the token renewal interval is at least hours in practice. If I'm right here, it doesn't sound a crazy idea to always call the token.renew() to get the "correct" next renewal dates and pick the minimum value. This would eliminate almost all the safeguards and simplifies the logic greatly, having trade-off Spark should make calls per each token update.

Does this make sense?

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35642/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35642/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/35642/
Test FAILed.

SparkQA commented:
**[Test build #131036 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131036/testReport)** for PR 30366 at commit [`38a6c5a`](https://github.com/apache/spark/commit/38a6c5a7232de90f41cc00342173ae41884fb9ad).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131036/
Test PASSed.

tgravescs commented:
@HeartSaVioR thanks for the detailed description and analysis, that is very helpful.  
I thought the way we did it was just got the earliest renewal so we didn't have to track all them individually because the common case is renewal only happens in hours time frame - like once every 24 hours but its obviously configurable. At that point you just check all of them. I'll have to look at the code in more detail to make sure I remember correctly and I think things have changed with code refactoring that I didn't review.   
In general it seems bad for any token to have renewal that is under a certain threshold, definitely shouldn't be immediate all the time.  I'll look some more this afternoon.

dongjoon-hyun reviewed (commented):
Thanks, @HeartSaVioR . BTW, it would be great if we can have a test case with mocking.

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/deploy/security/HadoopFSDelegationTokenProvider.scala:
```diff
@@ -50,25 +51,34 @@ private[deploy] class HadoopFSDelegationTokenProvider
+          val nextRenewalDateForToken = issueDate + interval
+          logDebug(s"Next renewal date is $nextRenewalDateForToken for token ${kind.toString}")
+          nextRenewalDateForToken
+        }.filterNot(_ < currentTime)
```
Can we hit the situation where `_ < currentTime` for all tokens? Do we need to distinguish this case from the case `tokenRenewalIntervals` is empty?

dongjoon-hyun commented on core/src/main/scala/org/apache/spark/deploy/security/HadoopDelegationTokenManager.scala:
```diff
@@ -230,6 +230,8 @@ private[spark] class HadoopDelegationTokenManager(
         val ratio = sparkConf.get(CREDENTIALS_RENEWAL_INTERVAL_RATIO)
         val delay = (ratio * (nextRenewal - now)).toLong
+        logInfo(s"Calculated delay on renewal is $delay, based on next renewal $nextRenewal " +
+          s"and the ratio $ratio")
```
Although the log message has timestamp, it would be good to have `now` information together in addition to `nextRenewal` in this *log* line because it will be subtracted and can be *minus*.

HeartSaVioR commented:
@tgravescs 

Thanks for the comment.

> I thought the way we did it was just got the earliest renewal so we didn't have to track all them individually because the common case is renewal only happens in hours time frame - like once every 24 hours but its obviously configurable. 

Yes that's configurable but agree that it only happens in hours time frame. The problematic token also produced expired time as 7 days later, so even longer than the Hadoop delegation token. That is simply just because Spark has a belief that token identifier should have valid issue date, whereas it's not "guaranteed" for every implementations. Once the precondition is broken, the calculation is completely going on unintentional way. There's safeguard but the safeguard also makes thing bad.

That's why I had to change the approach. Before reaching this approach, I just fixed the issue via simply discarding the invalid next renewal date like less than now, but that really feels me just adding a workaround. (This workaround also can't handle the new case I left in above comment.) And given we're here, I'd like to collect some ideas to make this "concrete" with tolerating known case of broken assumption.

HeartSaVioR commented:
@dongjoon-hyun 
Thanks for the review. Given @tgravescs is revisiting the logic (and considering the above comments as well), I expect the code change would occur afterwards. I'd defer addressing review comments until we are in agreement with the direction and the PR reflects the consensus.

In the meanwhile I'll try to look into how to add the test.

HeartSaVioR commented:
I just changed the PR to draft, as I would make sure we consider about the better change (than the current approach of the PR) accounting my comments above.

Reviewers, please consider the code change to only the reference, and go through below comments and share your opinion. Thanks in advance!

https://github.com/apache/spark/pull/30366#issuecomment-726531257
https://github.com/apache/spark/pull/30366#issuecomment-726534500

HeartSaVioR commented:
I see another problematic case where there're two token identifiers (say, token A and B) which token B's identifier provides issue date as 0L (same as the case provided in PR description) but the token B requires shorter token renewal period than token A. In this case, we can't simply ignore the renewal period calculation for token B.

And I'm also in doubt with caching the renewal interval. Suppose the same case, but token A has intermittent issue and ends up throwing exception when token.renew() is called for token A. This is "swallowed" without any log message and the renewal interval for the token A is never reconsidered.
(Probably we need to retry the call after some period, but the period should be also taken carefully in case of the token which token.renew() always throws exception. A bit complicated to also account for now?)

gaborgsomogyi commented:
AFAIU the problem is that the valid issue date is not filled. I think this is a bug because the whole concept is based on token provider voting. All providers give back the time when it requires renewal and the security manager takes a lowest one. This means couple of providers obtain tokens a little bit earlier but I think this shouldn't be change.

Just a question: Without valid issue date how should the security manager know when exactly the token need to be renewed?

HeartSaVioR commented:
Looks like Spark requires two preconditions to make scheduling renewal work correctly:

1) `token.renew()` should give valid expiration date, or should throw exception if it's not supported.
2) token identifier should give valid issue date.

This is from the case where only 1) is fulfilled. In this case, we can still schedule based on the "absolute" renewal date from `token.renew()`, but no longer be able to deal with "interval" which is relative to "issue date".

gaborgsomogyi commented:
I've had a deeper look and the title was a little bit misleading to me. I thought we intend to change the framework itself bit is seems like has effect only on HadoopFS provider. This makes the situation less invasive because the whole logic stays the same but some provider internals what we need to change. Though I think it would be include `HadoopFSDelegationTokenProvider` in the title.

I think not filling such fields is a bug in the client itself but I feel the pain from user perspective.
Without super deep consideration and since we're speaking about at least hours or days not yet sure why we're not just using `now` as issue date (with a super big warning printout)?
I need to dig a little bit deeper to have good internals knowledge (I've mainly reviewed and developer the framework itself).

HeartSaVioR commented:
Ah yes you're right. I should have clarified the boundary of change - HadoopFS provider is only affected. Let me add the information in PR title.

Regarding considering `now` as `issue date`, I agree that is probably one of easiest approaches and not hurting much, if there's a guarantee that fetchDelegationToken always update all tokens. I'm not expert here, so don't know about the case the precondition can be broken. That's why I'm hesitate to make a change based on that.

Btw, we're also opening the possibility to fix this in Hadoop side, as the problematic identifier was from KNOX and it inherited the abstract class in hadoop-aws. Once we fix it we could just keep the precondition that issue date should be valid. I'll try to deal with Hadoop side first, and revisit afterwards.

HeartSaVioR commented:
FYI, fixing S3 identifier's invalid issue date is tracked via https://issues.apache.org/jira/browse/HADOOP-17379

The question remains, would we still want to make the code be resilient to the bad token identifier, as there're existing Hadoop versions which contain the bug? 

If the answer is yes, we could weigh the options we enumerated. IMHO, `consider the issue date as now` is the simplest one, just not 100% sure it's safe to assume.

If the answer is no, we can close this PR and mark the issue as "Won't track here" or likewise.

gaborgsomogyi commented:
Good to hear the bug will be fixed on Hadoop side.

> The question remains, would we still want to make the code be resilient to the bad token identifier, as there're existing Hadoop versions which contain the bug?

My personal opinion is yes and no. I don't think we should solve all problem what token providers generate especially because the change can be either risky (using `now`) or complex which could bring in new issues (tracking individually).
What I would do is writing out an error message what causes the issue not to spend huge amount of time to find it out next time.

tgravescs commented:
sorry for my delay, I got busy with other things. Yeah fixing the Hadoop side is great. Its been a long time since I looked at tokens so tried to look at the Hadoop code to refresh my memory.  

I would say it doesn't hurt for us to protect if the issue date is 0 by using now and then logging a warning that we are doing so. That is what I thought the Hadoop secret managers did with it on renewal.

I'm assuming the issueDate never changed from 0?  meaning it started 0 and even after renewal it stayed 0, right?  Do you know what class was renewing?

gaborgsomogyi commented:
> I would say it doesn't hurt for us to protect if the issue date is 0 by using now and then logging a warning that we are doing so. That is what I thought the Hadoop secret managers did with it on renewal.

Could you point some Hadoop code to understand your point?

tgravescs commented:
https://github.com/apache/hadoop/blob/trunk/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/token/delegation/AbstractDelegationTokenSecretManager.java#L423

gaborgsomogyi commented:
Do we re-obtain these tokens or renew already existing as well? `Now` goes on when we renew.

HeartSaVioR commented:
https://github.com/apache/hadoop/commit/22039a14ff2b9d280f71b0735c32e6ac717e2344

Please take a look at the commit message body. `AbstractDelegationTokenIdentifier` has issue date and max date as 0L by default, and being assigned when `AbstractDelegationTokenSecretManager.createPassword` is called. That said, any token identifiers which don't follow the path should initialize issue date and max date by themselves.

`AbstractS3ATokenIdentifier` extends `AbstractDelegationTokenIdentifier`, but includes `getCreated()` and `getExpiryTime()` as well which look to conflict with issue date and max date. Implementations of `AbstractS3ATokenIdentifier` may try to provide the correct values for both methods, but at the risk of not updating issue date and max date.

I actually encountered this from internal class (in our org), but we just made sure at least issue date is valid for `AbstractS3ATokenIdentifier` by HADOOP-17379.

HeartSaVioR commented:
cc. @steveloughran 
Could you please provide an input whether it would work if we consider `issue date` as `current timestamp` when `issue date` is not properly initialized (0L)? That's based on precondition that we always fetch the delegation tokens just before the calculation is happening. Just to get some help from experts so that we don't miss anything.

steveloughran commented:
Classic HDFS tokens can be renewed generally up to a week and also revoked. The S3ADT is which we ship with Apache Hadoop are not renewable at all… Which is why I hadn't noticed I was getting this wrong. The Knox DTs are renewable as presumably, are those which HD/Insights supports. I don't know if they are doing the right thing. One recommendation could be to actually print a warning if a token has a an issue date of zero.

As to what makes a good default, the usual use case of tokens collected in Spark application launch using now as the default issue date seems absolutely perfect. It is only in the special case of tokens collected in advance and the token implementation doesn't set issue date that there would be any problem from this.

gaborgsomogyi commented:
Considering Steve's addition + in the meantime I've had a deeper look at the logic. I'm fine with `now` + a warning message.

HeartSaVioR commented:
OK cool. I'll update the PR, and remove WIP. As we will simply change the issue date I don't think new test is required with the new change.

HeartSaVioR committed:
Just simplify the approach as discussed

SparkQA commented:
**[Test build #131301 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131301/testReport)** for PR 30366 at commit [`59d7d40`](https://github.com/apache/spark/commit/59d7d40f81b1728a99546869f585b46f62f7eea2).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35905/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35905/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/35905/
Test PASSed.

dongjoon-hyun reviewed (dismissed):
Sorry, but why do you think we don't need a new test? Is it due to the difficulty?
> As we will simply change the issue date I don't think new test is required with the new change.

SparkQA commented:
**[Test build #131301 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131301/testReport)** for PR 30366 at commit [`59d7d40`](https://github.com/apache/spark/commit/59d7d40f81b1728a99546869f585b46f62f7eea2).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131301/
Test FAILed.

HeartSaVioR commented:
> Sorry, but why do you think we don't need a new test? Is it due to the difficulty?

Yes. This PR only brings 30 lines of changes which is enough intuitive to be reviewed easily, whereas adding a new test incurs mocking everything including token and token identifier which is even harder than the original fix.

This is a sort of "broken window" - the complicated logic is already introduced without any unit tests, and we are trying to enforce adding test from scratch and paying off the old debt instead. I'd be appreciated if you can volunteer to pay off the debt on existing code logic, and then I'll craft a test based on that. Otherwise I'd wonder this is something I should do; time is precious, for everyone.

gaborgsomogyi commented on core/src/main/scala/org/apache/spark/deploy/security/HadoopFSDelegationTokenProvider.scala:
```diff
@@ -126,13 +130,28 @@ private[deploy] class HadoopFSDelegationTokenProvider
+    if (identifier.getIssueDate > 0L) {
+      identifier.getIssueDate
+    } else {
+      if (!tokensWarnedIssueDateProblem.contains(kind)) {
```
I think we need to write it out all the time (this happens maybe once a day). Several times I'm getting partial logs which makes it hard to guess.

HeartSaVioR replied:
Makes sense. At least we wouldn't have flooding log message. Probably I was over-thinking.

gaborgsomogyi commented on core/src/main/scala/org/apache/spark/deploy/security/HadoopFSDelegationTokenProvider.scala:
```diff
@@ -126,13 +130,28 @@ private[deploy] class HadoopFSDelegationTokenProvider
+    } else {
+      if (!tokensWarnedIssueDateProblem.contains(kind)) {
+        logWarning(s"Token $kind has not set up issue date properly. Using current timestamp as" +
+          " issue date as a workaround. Consult token implementator to fix the behavior.")
```
I would write out the given issue date too.

gaborgsomogyi commented on core/src/main/scala/org/apache/spark/deploy/security/HadoopFSDelegationTokenProvider.scala:
```diff
@@ -126,13 +130,28 @@ private[deploy] class HadoopFSDelegationTokenProvider
     if (renewIntervals.isEmpty) None else Some(renewIntervals.min)
   }
+
+  private def getIssueDate(kind: String, identifier: AbstractDelegationTokenIdentifier): Long = {
```
Since others asked for tests maybe we can make this package private and add positive and negative cases (hitting both side of the if condition). I personally don't insist to this.

HeartSaVioR replied:
The reason I was negative on adding test is, as we know how thing was broken, we basically expect the test to simulate it and verify it's no longer failing. e.g. testing `obtainDelegationTokens()`. `obtainDelegationTokens()` calls various private methods which call Hadoop API, hence requiring mocks for everything. This is actually an existing debt, and non-trivial to pay off, compared to what I did for the fix.

I'm totally OK to add test verifying this method, as it only requires me to mock AbstractDelegationTokenIdentifier. I'm not sure this fits the needs of adding test, though.

gaborgsomogyi replied:
That said I don't insist on the test because I agree with you. Just wanted to come up with some coverage. Creating a normal test with security manager + KDC + providers + hadoop sever would require a brain surgery complicated code (I've tried it before in the Kafka area but it was not a success story).

@dongjoon-hyun WDYT?

steveloughran replied:
You don't need a security manager for fetching DTs of filesystems or renewing. Example: the S3A ones can be issued if you have AWS creds. It's just that all code written today assumes that filesystems & services only issue DTs if user is kerberized or was launched with an existing set of creds. Example: how to collect a DT to talk to the landsat bucket for the next 12 hours

```
bin/hdfs fetchdt -D fs.s3a.delegation.token.binding=org.apache.hadoop.fs.s3a.auth.delegation.SessionTokenBinding --webservice s3a://landsat-pds/ secrets.bin
```

Renewal is a different bit of work tho'. It would get complicated fast

gaborgsomogyi replied:
What I've learned from past things is depending on external executables is not always a good decision. This tool must be available on all OS and must behave the exact same way otherwise the test will be pain in the ass.

gaborgsomogyi replied:
Maybe this is the case here but I'm always a little afraid when this pops up.

HeartSaVioR replied:
This discussion shows how hard to deal with test obtainDelegationTokens(). I hope any hero jumps in and deals with this.

Another approach would be changing all private methods to protected methods and fake via subclassing (say, not calling Hadoop API whenever possible), but we lose the ability to test the original methods we override, as well as we modify the code just because we need to add tests. A bit odd rationalization.

gaborgsomogyi replied:
There is a pretty good reason behind why only sub-parts of token providers are tested, please see Kafka for example...
I still say test getIssueDate.

gaborgsomogyi replied:
If you still need the pain maybe you can try to mock HadoopFSDelegationTokenProvider.hadoopFSsToAccess with powermock and generate a faulty token. That could lead far.

HeartSaVioR replied:
```
  private def getIssueDate(kind: String, identifier: AbstractDelegationTokenIdentifier): Long = {
    val issueDate = identifier.getIssueDate
    if (issueDate > 0L) {
      issueDate
    } else {
      val now = System.currentTimeMillis()
      logWarning(s"Token $kind has not set up issue date properly. (provided: $issueDate) " +
        s"Using current timestamp ($now) as issue date instead. Consult token implementor to fix " +
        "the behavior.")
      now
    }
  }
```

This doesn't seem to be something we would really want to have tests. Unless we separate out the calculation logic and try to craft tests, I don't see any option to have tests for this class.

gaborgsomogyi reviewed (commented):
Mainly looks good, some minors.

steveloughran reviewed (commented):
LGTM. Note we've got one more change due on hadoop-common to use UTC as the issue date.

This raises another question: what to do if the issue date > now? That is it thinks it was issued a few hours into the future? What happens then?

dongjoon-hyun commented:
According to the author's opinion, I dismissed my review comments. We can add a test case later if we need.

dongjoon-hyun commented:
Could you review this, @mridulm and @tgravescs ?

HeartSaVioR commented:
> LGTM. Note we've got one more change due on hadoop-common to use UTC as the issue date.
>
> This raises another question: what to do if the issue date > now? That is it thinks it was issued a few hours into the future? What happens then?

AFAIK we don't have any guard around that. If the result of `token.renew()` is later than issue date, we still get correct value of renew interval, but calculating next renewal date would be problematic as we get it via `issue date + renew interval`. 

If the renewal date goes to be less than now, we adjust the negative value as 0, and continuous scheduling without delay happens like I described in the PR description. If the renewal date goes to be higher than actual expire date, well, we know the result already.

HeartSaVioR committed:
Review comments

SparkQA commented:
**[Test build #131375 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131375/testReport)** for PR 30366 at commit [`bc472fd`](https://github.com/apache/spark/commit/bc472fd36def3a010eb507694c0899bf4be32faa).

SparkQA commented:
**[Test build #131375 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131375/testReport)** for PR 30366 at commit [`bc472fd`](https://github.com/apache/spark/commit/bc472fd36def3a010eb507694c0899bf4be32faa).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131375/
Test FAILed.

HeartSaVioR committed:
Remove unused import

SparkQA commented:
**[Test build #131376 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131376/testReport)** for PR 30366 at commit [`abbbc48`](https://github.com/apache/spark/commit/abbbc486960d27afb5fcd9ed822f909dbbb5b8fa).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35978/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35979/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35978/

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/35978/
Test PASSed.

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/35979/

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/35979/
Test FAILed.

SparkQA commented:
**[Test build #131376 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131376/testReport)** for PR 30366 at commit [`abbbc48`](https://github.com/apache/spark/commit/abbbc486960d27afb5fcd9ed822f909dbbb5b8fa).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131376/
Test PASSed.

HeartSaVioR commented:
cc. @tgravescs @mridulm Friendly reminder.

gaborgsomogyi reviewed (approved):
LGTM from my side too. Adding test would be hell complicated (if possible).

steveloughran commented:
I think you should also log when the issueDate is > the time on the local machine. As discussed, fixing up is hard, but at least printing out "clock mismatch" is a useful thing to see in the logs when you hit problems

HeartSaVioR committed:
Add logging message for issue date > now

HeartSaVioR commented:
OK I think that's a good addition. Updated.

@tgravescs I'm sorry, but could you take a quick look again? Thanks in advance.

SparkQA commented:
**[Test build #131836 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131836/testReport)** for PR 30366 at commit [`90216ff`](https://github.com/apache/spark/commit/90216ff65f5a45d5906416d4f6000d4b3858b68a).

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36433/

steveloughran commented:
thanks; hopefully we won't ever see this in the logs, but at least when it causes renewal problems then there'll be something in the logs to point the blame at clock settings in whatever issued the token

SparkQA commented:
**[Test build #131836 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131836/testReport)** for PR 30366 at commit [`90216ff`](https://github.com/apache/spark/commit/90216ff65f5a45d5906416d4f6000d4b3858b68a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131836/

HeartSaVioR committed:
Remove unnecessary val

SparkQA commented:
**[Test build #131945 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131945/testReport)** for PR 30366 at commit [`b749574`](https://github.com/apache/spark/commit/b749574fb2b3d11037c4ff48a80910d63aa01455).

HeartSaVioR commented:
@tgravescs Friendly reminder.

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/36542/

SparkQA commented:
**[Test build #131945 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/131945/testReport)** for PR 30366 at commit [`b749574`](https://github.com/apache/spark/commit/b749574fb2b3d11037c4ff48a80910d63aa01455).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/131945/

tgravescs reviewed (approved):
adding the log message looks fine

HeartSaVioR commented:
Thanks all for reviewing! Merged into master and branch-3.0.
